### DISCUSSION TITLE
plot_predictions for a subset of groups

### DISCUSSION THREAD

**Discussion by andins** [bambinos/bambi#811] [Q&A] [answered]
I'm fitting a hierarchical model specified as:

```python
model = bmb.Model(
    "y ~ (1|group_id) + x1 + x2",
    data,
    family="t"
)
```

so I have two fixed regressors and the intercept varying by group.
Now I'd like to visualize model predictions (together with the data) as a function of x1 for a subset of groups (in different panels).
Is it possible to do this using interpret.plot_predictions?
I achieved to plot the predictions for all of the groups but this is a large number (>300) so I'd rather look at a specific subset.
Please let me know if more details are needed.
Thank you in advance.

**Comment by GStechschulte** [comment, marked as answer]
Hey @andins thanks for raising the discussion. Yes, this should be possible. You would want to pass a specific subset of `group_id` values to `conditional`.  Optionally, you could also pass `x1` and `x2`. Otherwise, if you don't pass terms to `conditional` that were in the model, then we compute default values.

I imagine something like the following

```python
bmb.interpret.plot_predictions(
model=...
idata=...
conditional={ 
   "group_id: [50, 51, 52, 53, 54, 55, 150, 151, 152, 153, 154, 155],
   "x1": np.linspace(x1.min(), x1.max(), 50),
  "x2": np.mean(x2)
),
subplot_kwargs={"main": "x1", "group": "group_id", "panel": "group_id"}
)
```

which would plot a panel for each `group_id`.

**Reply by andins** [reply]
Hi Gabriel,

Thank you so much for your answer!
The approach you suggest works beautifully!
Trying to understand it a bit more I run it with a simulated dataset and noticed something strange.
If the groups that I pass to the conditional parameter in plot_predictions are not sorted, they get sorted internally and the returned axes will correspond to the sorted groups.
Is this the intended behavior? I might be interested in sorting the groups by e.g. group intercept or group effect for x1.
I noticed this because the predictions were not making sense with the data when the groups were not sorted.
Below you have some code to reproduce this, if it helps.
Simulated data is just groups of gaussians.
I'm using numpyro for speed but the same happens using default.
Please let me know if anything is unclear of you need more info.
Together with the prediction and the data I'm also plotting the regression line from mean parameter samples as a reference.
Sorting variable plot_groups will fix the misalignment but then I'm not able to enforce a given order of groups.
```
import pandas as pd
import numpy as np
import bambi as bmb
import matplotlib.pyplot as plt

def sim_data(groups):
    n_groups = len(groups)
    data = pd.DataFrame()
    for g in groups:
        n = np.random.randint(4, 50)
        x1 = np.random.normal(loc=50, scale=20, size=n)
        x2 = np.random.normal(loc=20, scale=3, size=n)
        alpha = np.random.normal(0, 1)
        beta1 = np.random.normal(loc=2, scale=1)
        beta2 = np.random.normal(loc=1, scale=1)
        y_scale = 10
        y = np.random.normal(loc=alpha + beta1 * x1 + beta2 * x2, scale=y_scale, size=n)
        data = pd.concat(
            (data, pd.DataFrame({"group_id": g, "n": n, "x1": x1, "x2": x2, "y": y}))
        )
    return data

def abline(slope, intercept, ax=None, *args, **kwargs):
    if ax is None:
        ax = plt.gca()
    x_vals = np.array(ax.get_xlim())
    y_vals = intercept + slope * x_vals
    ax.plot(x_vals, y_vals, *args, **kwargs)
    return ax

def plot_predictions_with_data(
    model,
    idata,
    data,
    conditional,
    subplot_kwargs,
    fig_kwargs,
    y_var="y",
    group_var="group_id",
    x_var="x1",
    additional_covariates=[],
):
    fig, axs = bmb.interpret.plot_predictions(
        model=model,
        idata=idata,
        conditional=conditional,
        subplot_kwargs=subplot_kwargs,
        fig_kwargs=fig_kwargs,
    )

    axes_flat = axs.ravel()
    for i, p in enumerate(conditional[group_var]):
        ax = axes_flat[i]
        zeroth_idx = np.where(data[group_var] == p)[0]

        x_values = data.iloc[zeroth_idx][x_var].values
        y_values = data.iloc[zeroth_idx][y_var].values
        ax.scatter(x_values, y_values)
        # add regression line
        intercept = (
            idata["posterior"][f"1|{group_var}"]
            .where(
                idata["posterior"][f"1|{group_var}"][f"{group_var}__factor_dim"] == p,
                drop=True,
            )
            .mean()
            .values
        )
        additional_cov_intercept = 0
        for x in additional_covariates:
            additional_cov_intercept += (
                idata["posterior"][x].mean().values * data[x].mean()
            )
        intercept += additional_cov_intercept
        slope = (
            idata["posterior"][f"{x_var}|{group_var}"]
            .where(
                idata["posterior"][f"{x_var}|{group_var}"][f"{group_var}__factor_dim"] == p,
                drop=True,
            )
            .mean()
            .values
        )
        abline(slope, intercept, ax=ax, c="purple", linestyle="--")
        ######################
        ax.set_xlabel("x1")
        ax.set_ylabel("y")
        ax.set_title(f"{group_var} {p}")
        ax.get_legend().remove()

    plt.tight_layout() 
    return fig, ax

# if the groups are not sorted plot_predictions is misaligned
data = sim_data(["b", "d", "a", "c", "e", "f", "g", "h", "i", "j"])


model_synth = bmb.Model(
    "y ~ (1 + x1|group_id) + 0 + x2",
    data,
    family="gaussian",
)
results_synth = model_synth.fit(
    draws=1000, inference_method="nuts_numpyro", target_accept=0.96
)

plot_groups = data.group_id.unique()
plot_predictions_with_data(
    model_synth,
    results_synth,
    data,
    conditional={
        "group_id": plot_groups,
        "x1": np.linspace(data.x1.min(), data.x1.max(), 50),
        "x2": np.mean(data.x2),
    },
    subplot_kwargs={"main": "x1", "group": "group_id", "panel": "group_id"},
    fig_kwargs=dict(figsize=(11, 8)),
    additional_covariates=["x2"],
)
```

When I run this I get something like this
![image](https://github.com/bambinos/bambi/assets/12953365/c807fea6-4012-49de-8256-5b397cca5fb7)

**Reply by GStechschulte** [reply]
Many thanks for providing a reproducible example. This is interesting. I can't necessarily say anything off the top of my head, but I will have a look at it within the coming days. Thanks! :)

**Reply by GStechschulte** [reply]
Hey @andins apologies for the delayed response. 

> If the groups that I pass to the conditional parameter in plot_predictions are not sorted, they get sorted internally and the returned axes will correspond to the sorted groups. Is this the intended behavior? 

When plotting a `group` and `panel` covariate, Bambi [computes](https://github.com/bambinos/bambi/blob/f36f4345d8fe36cdb1129b42900ce5824bbab8b7/bambi/interpret/plot_types.py#L94) the unique values for these covariates before plotting. The default behavior of `np.unique` is to return a sorted array.

If I plot your data not using your function, it appears everything is working fine.

```python3
bmb.interpret.plot_predictions(
    model_synth,
    results_synth,
    conditional={
        "group_id": data.group_id.unique(),
        "x1": np.linspace(data.x1.min(), data.x1.max(), 50),
        "x2": np.mean(data.x2),
    },
    subplot_kwargs={"main": "x1", "group": "group_id", "panel": "group_id"},
    fig_kwargs=dict(figsize=(11, 8)),
    legend=False
)
plt.tight_layout();
```
![image](https://github.com/bambinos/bambi/assets/63432018/ca98a051-2b41-4c01-9f4b-56629cf63fba)

This discrepancy is due to the loop `for i, p in enumerate(conditional["group_id"]):` in your function which iterates over an unsorted array. If you sort and then iterate, your problem will be resolved.

```python3
conditional = {
        "group_id": data.group_id.unique(),
        "x1": np.linspace(data.x1.min(), data.x1.max(), 50),
        "x2": np.mean(data.x2),
    }

fig, ax = bmb.interpret.plot_predictions(
    model_synth,
    results_synth,
    conditional=conditional,
    subplot_kwargs={"main": "x1", "group": "group_id", "panel": "group_id"},
    fig_kwargs=dict(figsize=(11, 8)),
    legend=False
)

axes_flat = ax.ravel()
for i, p in enumerate(sorted(conditional["group_id"])):
    ax = axes_flat[i]
    zeroth_idx = np.where(data["group_id"] == p)[0]

    x_values = data.iloc[zeroth_idx]["x1"].values
    y_values = data.iloc[zeroth_idx]["y"].values
    ax.scatter(x_values, y_values)

plt.tight_layout()
```
![image](https://github.com/bambinos/bambi/assets/63432018/7609bb16-c888-469b-913a-74538633a224)

**Reply by andins** [reply]
Thank you very much for your answer!
It makes sense and it's what I suspected: I noticed that I could just sort the groups and get the right pairing of regression plots and data scatter.
As a follow up, is there a way to use enforce a different sorting of covariates in plot_predictions? Not in the toy example above but in my real use case I'm trying to plot groups in order of increasing intercept. Maybe the only way is to change the plot_predictions code...

**Reply by GStechschulte** [reply]
Ahh cool, good to hear. 

> is there a way to use enforce a different sorting of covariates in plot_predictions?

Not at the moment. If more users start to request this feature then I will look into this. An alternative way is to: (1) use `bmb.interpret.predictions` which returns a dataframe, then plot this dataframe how you wish, or (2) follow the [advanced interpret usage](https://bambinos.github.io/bambi/notebooks/interpret_advanced_usage.html) to access lower level interpret functions.

**Reply by andins** [reply]
It makes sense. Thank you very much for the suggestion!
